nextcord/cache.py

```diff
@@ -0,0 +1,149 @@
+__get_item__ = dict.__getitem__
+__get__ = dict.get
+__pop__ = dict.pop
+
+
+class CacheMissing(Enum):
+    MISSING = 0
+
+
+_MISSING = CacheMissing.MISSING
+
+
+class SizedDict(dict):
+    """This represents a sized cache dictionary where objects over size limit
+    will be removed upon adding a new object.
+
+    .. note::
+        This cache relies on the presumed order of dictionaries after Python 3.7
+        This allows for the cache to be much faster and take 3x less memory than
+        the TTL cache from the `collections` module.
+    """
+
+    __slots__ = ("maxsize",)
+
+    def __init__(self, maxsize: Optional[Union[float, int]] = None) -> None:
+        self.maxsize: Union[float, int] = maxsize or float("inf")
+
+    def __setitem__(self, key, val) -> None:
+        """Ensures doesn't exceed max size before adding new item."""
+        if len(self) >= self.maxsize:
+            for item in self:  # iter to delete first item
+                del self[item]
+                break
+        __set_item__(self, key, val)
+
+
+class BetterTTLCache(dict):
+    """This represents a timed cache dictionary where objects past a certain age
+    will be removed upon the next get (lazy ttl), and objects over size limit
+    will be removed upon setting a new object.
+
+    .. note::
+        This cache relies on the presumed order of dictionaries after Python 3.7
+        This allows for the cache to be much faster and take 3x less memory than
+        the TTL cache from the `collections` module.
+    """
+
+    __slots__ = ("maxsize", "ttl", "timelink", "working")
+
+    def __init__(
+        self, maxsize: Optional[int] = None, ttl: Optional[Union[int, float]] = None
+    ) -> None:
+        self.maxsize: Union[int, float] = maxsize or float("inf")
+        self.ttl: Optional[Union[int, float]] = ttl
+        self.timelink: dict[Hashable, float] = {}
+        self.working: bool = False
+
+    def __getitem__(self, key):
+        """Expires items before returning getitem."""
+        self.__expire__()
+        return __get_item__(self, key)
+
+    def __setitem__(self, key, val) -> None:
+        """
+        Expires and ensures self max size is maintained before setting new item.
+        """
+        self.__expire__()
+        if len(self) >= self.maxsize:
+            for item in self:  # iter to delete first item
+                del self[item]
+                break
+        __set_item__(self, key, val)
+        self.update_time(key)
+
+    def __expire__(self) -> None:
+        """
+        Takes care of expiring keys in the timelink that have passed their ttl lifespan.
+        Uses a rolling motion for expiring, all items at the start of the timelink
+        are assumed to be the earliest created items to avoid checking every item.
+        Timelink order must be managed properly in order for expiration to work.
+        """
+        if not self.ttl or self.working:
+            return
+
+        self.working = True

```

Comment:
I'm confused what this is for? Threading support?

Reply:
I'm not sure whether it's necessary to be honest, I had it as an extra layer of insurance. Ensures the method doesn't get called twice at once, if that happens it can cause an error during iteration.

Reply:
Due to no async calls being made and ignoring threading (we discourage threading), `__expire__` shouldn't ever be able to run while it's already running.

Reply:
Then I could remove this for a tiny speedup, would it be worth it in case the user decides to access the cache from a thread and it completely breaks the ttl structure?

Reply:
If threading support was important, we'd want to use something like `threading.Lock` to properly make sure we don't have multiple `__expire__`'s running.
Also as is, iterating through `BetterTTLCache().items()` and triggering `__expire__` would cause a Changed Size During Iteration error if `__expire__` hit the `del self[item]` stuff. I think you'd need to have everything use the same lock to prevent that, and tbh throwing a "Does not support threading" docstring in the class seems easier than doing all of that.

Reply:
Fair, I will add that next commit